power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/dataset.hpp

```diff
@@ -0,0 +1,170 @@
+        return total_elements;
+    }
+    Idx component_total_elements(Idx component_idx) const { return component_total_elements(*this, component_idx); }
+
+  private:
+    Handle handle_{};
+    std::unique_ptr<PGM_DatasetInfo const> info_;
+};
+
+class DatasetConst {
+  public:
+    DatasetConst(std::string const& dataset, Idx is_batch, Idx batch_size)
+        : dataset_{PGM_create_dataset_const(handle_.get(), dataset.c_str(), is_batch, batch_size)} {}
+    DatasetConst(DatasetWritable const& writable_dataset)
+        : dataset_{PGM_create_dataset_const_from_writable(handle_.get(), writable_dataset.get())} {}
+    DatasetConst(DatasetMutable const& mutable_dataset)
+        : dataset_{PGM_create_dataset_const_from_mutable(handle_.get(), mutable_dataset.get())} {}
+
+    RawConstDataset* get() const { return dataset_.get(); }
+
+    static void add_buffer(DatasetConst& dataset, std::string const& component, Idx elements_per_scenario,
+                           Idx total_elements, Idx const* indptr, RawDataConstPtr data) {
+        PGM_dataset_const_add_buffer(dataset.handle_.get(), dataset.dataset_.get(), component.c_str(),
+                                     elements_per_scenario, total_elements, indptr, data);
+        dataset.handle_.check_error();
+    }
+    void add_buffer(std::string const& component, Idx elements_per_scenario, Idx total_elements, Idx const* indptr,
+                    RawDataConstPtr data) {
+        return add_buffer(*this, component, elements_per_scenario, total_elements, indptr, data);
+    }
+
+    static Info const get_info(DatasetConst const& dataset) {
+        auto const info = PGM_dataset_const_get_info(dataset.handle_.get(), dataset.dataset_.get());
+        dataset.handle_.check_error();
+        return Info(info);
+    }
+    Info const get_info() const { return get_info(*this); }
+
+  private:
+    Handle handle_{};
+    detail::UniquePtr<RawConstDataset, PGM_destroy_dataset_const> dataset_;
+};
+
+class DatasetWritable {
+  public:
+    RawWritableDataset* get() const { return dataset_.get(); }
+
+    static Info const get_info(DatasetWritable const& dataset) {
+        auto const info = PGM_dataset_writable_get_info(dataset.handle_.get(), dataset.dataset_.get());
+        dataset.handle_.check_error();
+        return Info(info);
+    }
+    Info const get_info() const { return get_info(*this); }
+
+    static void set_buffer(DatasetWritable const& dataset, std::string const& component, Idx* indptr, RawDataPtr data) {
+        PGM_dataset_writable_set_buffer(dataset.handle_.get(), dataset.dataset_.get(), component.c_str(), indptr, data);
+        dataset.handle_.check_error();
+    }
+    void set_buffer(std::string const& component, Idx* indptr, RawDataPtr data) {
+        return set_buffer(*this, component.c_str(), indptr, data);
+    }
+
+  private:
+    friend class Deserializer;
+    DatasetWritable(RawWritableDataset* dataset) : dataset_{dataset} {}
+    Handle handle_{};
```

Comment:
make sure that `handle_` always defined at the first class member. You need this for other stuff.

Reply:
Silly mistake. Done.